pandas/core/indexes/base.py

```diff
@@ -2796,14 +2796,18 @@ def intersection(self, other, sort=False):
         """
         self._validate_sort_keyword(sort)
         self._assert_can_do_setop(other)
-        other, _ = self._convert_can_do_setop(other)
+        other, result_name = self._convert_can_do_setop(other)
 
         if self.equals(other):
             if self.has_duplicates:
                 return self.unique()._get_reconciled_name_object(other)
             return self._get_reconciled_name_object(other)
 
-        if not is_dtype_equal(self.dtype, other.dtype):
+        elif not self._should_compare(other):
```

Comment:
we have test cases for this right? this is just a short-circuit

Reply:
yes, about a dozen each